EIPS/eip-2537.md

```diff
@@ -0,0 +1,293 @@
+---
+eip: 2537
+title: BLS12-381 curve operations
+author: Alex Vlasov (@shamatar)
+discussions-to: https://github.com/ethereum/EIPs/pull/2537
+status: Draft
+type: Standards Track
+category: Core
+created: 2020-02-21
+requires : 1109, 2046
+---
+
+<!--You can leave these HTML comments in your merged EIP and delete the visible duplicate text guides, they will not appear and may be helpful to refer to if you edit it again. This is the suggested template for new EIPs. Note that an EIP number will be assigned by an editor. When opening a pull request to submit your EIP, please use an abbreviated title in the filename, `eip-draft_title_abbrev.md`. The title should be 44 characters or less.-->
+
+## Simple Summary
+<!--"If you can't explain it simply, you don't understand it well enough." Provide a simplified and layman-accessible explanation of the EIP.-->
+
+This precompile adds operation on BLS12-381 curve as a precompile in a set necessary to *efficiently* perform operations such as BLS signature verification and perform SNARKs verifications.
+
+## Abstract
+<!--A short (~200 word) description of the technical issue being addressed.-->
+
+If `block.number >= X` we introduce *eight* separate precompiles to perform the following operations (addresses to be determined) with one of them being optional and up to discussion (BLS12-381 curve allows efficient square root extraction, so point decompression makes *some* sense):
+
+- G1ADD - to perform point addition on a curve defined over prime field
+- G1MUL - to perform point multiplication on a curve defined over prime field
+- G1MULTIEXP - to perform multiexponentiation on a curve defined over prime field
+- G2ADD - to perform point addition on a curve twist defined over quadratic extension of the base field
+- G2MUL - to perform point multiplication on a curve twist defined over quadratic extension of the base field
+- G2MULTIEXP - to perform multiexponentiation on a curve twist defined over quadratic extension of the base field
+- PAIRING - to perform a pairing operations between a set of *pairs* of (G1, G2) points
+- MAP_FIELD_TO_CURVE - maps base field element or extension field element into the curve point (G1 or G2 respectively) according to [IEFT specification](https://github.com/cfrg/draft-irtf-cfrg-hash-to-curve/blob/master/draft-irtf-cfrg-hash-to-curve.md#deterministic-mappings-mappings) using an simplified SWU method. It does NOT perform mapping of the byte string into field element that can be implemented in many different ways and can be efficiently performed in EVM, but only does field arithmetic to map field element into curve point. Such functionality is required for signature schemes.
```

Comment:
It seems this functionality implements an IETF draft. Are we at risk codifying a precompile which turns out to be not useful if the final version is sufficiently different?

Reply:
There is always such risk, but this has already happened with keccak256 vs SHA3. Worst case we update the functionality of the precompile or make a new one with new set of parameters. Mapping algorithms itself and isogeny transformation are universal and I would only expect parameters to change. Note that for universality reason I didn’t include a `3 mod 4` optimization into the spec cause we would later want the similar things for other curves.

Reply:
> Worst case we update the functionality of the precompile or make a new one with new set of parameters.

I'm not sure it is possible to update (== change the logic) a precompile once it has been deployed. This would mean introduction a new precompile for the final spec and leave the old one "to rot".

Reply:
So far it's the best we can get. Mapping of the field element to curve may change if more optimal ways are found, and then it'll be a legacy solution.

Reply:
cc @ralexstokes @vbuterin -- so is 2.0 based on the IETF draft? Is this a path already taken at this point?

Reply:
yep eth2 is following the IETF draft at this point. the idea was to coalesce efforts from various blockchain projects interested in BLS around this standardization process which we did.

no one wants another keccak vs sha3 for ethereum however we also don't want to delay the eth2 launch on finalizing the draft. it does seem like the parts that are still changing at this point we can do in the EVM, and the parts that have been more stable (and are resource intensive) are capture in the precompiles in this EIP